File: js/src/arguments.js
import Op from './evm/opcodes.js'
import { Vm, UnsupportedOpError } from './evm/vm.js'
import { StackIndexError } from './evm/stack.js'
import Element from './evm/element.js'
import { bigIntToUint8Array, uint8ArrayToBigInt, bigIntBitLength, toUint8Array } from './utils.js'

class Arg {
  offset
  path
  add_val
  and_mask
  constructor(properties) {
    Object.preventExtensions(this)
    Object.assign(this, properties)
  }
  toString() {
    return `Arg(off=${this.offset},path=${this.path},add_val=${this.add_val},and_mask=${this.and_mask})`
  }
}

class IsZeroResult {
  offset
  path
  add_val
  and_mask
  constructor(properties) {
    Object.preventExtensions(this)
    Object.assign(this, properties)
  }
  toString() {
    return `IsZeroResult(off=${this.offset},path=${this.path},add_val=${this.add_val},and_mask=${this.and_mask})`
  }
}

class InfoValDynamic {
  n_elements
  constructor(n) {
    this.n_elements = n
  }
  toString() {
    return `Dynamic(${this.n_elements})`
  }
}

class InfoValArray {
  n_elements
  constructor(n) {
    this.n_elements = n
  }
  toString() {
    return `Array(${this.n_elements})`
  }
}

class Info {
  constructor() {
    this.tinfo = null
    this.tname = null
    this.children = new Map()
  }
  toStr(isRoot = false) {
    if (this.tname !== null) {
      const [name] = this.tname
      if (name === 'bytes') {
        if (
          this.tinfo === null ||
          (this.tinfo instanceof InfoValArray && this.tinfo.n_elements === 0) ||
          (this.tinfo instanceof InfoValDynamic && this.tinfo.n_elements === 1)
        ) {
          return name
        }
      } else if (this.children.size === 0) {
        if (this.tinfo === null || this.tinfo instanceof InfoValDynamic) {
          return name
        }
      }
    }
    let startKey = this.tinfo instanceof InfoValArray ? 32 : 0
    let endKey = this.children.size > 0 ? Math.max(...this.children.keys()) : 0

    if (this.tinfo instanceof InfoValArray || this.tinfo instanceof InfoValDynamic) {
      endKey = Math.max(endKey, this.tinfo.n_elements * 32)
    }

    const q = []
    for (let k = startKey; k <= endKey; k += 32) {
      q.push(this.children.has(k) ? this.children.get(k).toStr(false) : 'uint256')
    }

    let c = q.length > 1 && !isRoot ? `(${q.join(',')})` : q.join(',')

    if (this.tinfo instanceof InfoValArray) {
      return `${c}[]`
    }
    if (this.tinfo instanceof InfoValDynamic) {
      if (endKey === 0 && this.children.size === 0) {
        return 'bytes'
      }
      if (endKey === 32) {
        if (this.children.size === 0) {
          return 'uint256[]'
        }
        if (this.children.size === 1 && this.children.values().next().value.tinfo === null) {
          return `${q[1]}[]`
        }
      }
    }
    return c
  }
}

class ArgsResult {
  constructor() {
    this.data = new Info()
    this.notBool = new Set()
  }

  getOrCreate(path) {
    return path.reduce((node, key) => {
      if (!node.children.has(key)) {
        node.children.set(key, new Info())
      }
      return node.children.get(key)
    }, this.data)
  }

  get(path) {
    return path.reduce((node, key) => {
      return node?.children.get(key)
    }, this.data)
  }

  markNotBool(path, offset) {
    const fullPath = [...path, offset]
    const el = this.get(fullPath)
    if (el && el.tname && el.tname[0] === 'bool') {
      el.tname = null
    }
    this.notBool.add(fullPath.join(','))
  }

  setTname(path, offset, tname, confidence) {
    const fullPath = offset !== null ? [...path, offset] : path
    if (tname === 'bool' && this.notBool.has(fullPath.join(','))) {
      return
    }
    const el = this.getOrCreate(fullPath)
    if (el.tname !== null && confidence <= el.tname[1]) {
      return
    }
    el.tname = [tname, confidence]
  }

  arrayInPath(path) {
    let el = this.data
    return path.map((p) => {
      if (el === undefined) return false
      el = el.children.get(p)
      return el && el.tinfo instanceof InfoValArray
    })
  }

  setInfo(path, tinfo) {
    if (path.length === 0) {
      // root
      return
    }

    const el = this.getOrCreate(path)
    if (tinfo instanceof InfoValDynamic) {
      if (el.tinfo instanceof InfoValDynamic && el.tinfo.n_elements > tinfo.n_elements) {
        return
      }
      if (el.tinfo instanceof InfoValArray) {
        return
      }
    }
    if (el.tinfo instanceof InfoValArray && tinfo instanceof InfoValArray) {
      if (tinfo.n_elements < el.tinfo.n_elements) {
        return
      }
    }
    el.tinfo = tinfo
  }

  joinToString() {
    return this.data.children.size === 0 ? '' : this.data.toStr(true)
  }
}

function andMaskToType(mask) {
  if (mask === 0n) {
    return null
  }
  if ((mask & (mask + 1n)) === 0n) {
    // 0x0000ffff
    const bl = bigIntBitLength(mask)
    if (bl % 8 === 0) {
      return bl === 160 ? 'address' : `uint${bl}`
    }
  } else {
    // 0xffff0000
    const m = BigInt(uint8ArrayToBigInt(bigIntToUint8Array(mask).slice().reverse()))
    if ((m & (m + 1n)) === 0n) {
      const bl = bigIntBitLength(m)
      if (bl % 8 == 0) {
        return `bytes${bl / 8}`
      }
    }
  }
  return null
}

export function functionArguments(code, selector, gas_limit = 5e4) {
  const code_arr = toUint8Array(code)
  const selector_arr = toUint8Array(selector)
  const vm = new Vm(code_arr, new Element(selector_arr, 'calldata'))

  let gas_used = 0
  let inside_function = false
  let args = new ArgsResult()

  while (!vm.stopped) {
    let ret
    try {
      if (inside_function) {
        // console.log('args:', args.joinToString());
        // console.log('not_bool:', args.notBool);
        // console.dir(args.data, { depth: null, colors: true })
        // console.log(vm.toString())
      }
      ret = vm.step()
      gas_used += ret[1]
      if (gas_used > gas_limit) {
        // throw `gas overflow: ${gas_used} > ${gas_limit}`
        break
      }
    } catch (e) {
      if (e instanceof StackIndexError || e instanceof UnsupportedOpError) {
        // console.log(e)
        break
      } else {
        throw e
      }
    }

    const [op, , r0, r1] = ret

    if (inside_function == false) {
      if (op === Op.EQ || op == Op.XOR || op == Op.SUB) {
        const p = vm.stack.peek().data[31]
        if (p === (op === Op.EQ ? 1 : 0)) {
          const a = r0.data.slice(-4)
          inside_function = selector_arr.every((v, i) => v === a[i])
        }
      }

      continue
    }

    switch (op) {
      case Op.CALLDATASIZE:
        vm.stack.pop()
        vm.stack.push_uint(131072n)
        break

      case Op.CALLDATALOAD:
      case Op.CALLDATACOPY:
        {
          if (r0.label instanceof Arg) {
            const { offset, path, add_val } = r0.label
            if (add_val >= 4 && (add_val - 4) % 32 === 0) {
              let po = 0
              if (add_val != 4) {
                const a = args.arrayInPath(path).reduce((s, v) => s + 32 * v, 0)
                if (a <= add_val - 4) {
                  po = a
                }
              }

              const fullPath = [...path, offset]
              const new_off = add_val - 4 - po

              args.setInfo(fullPath, new InfoValDynamic(new_off / 32))

              if (new_off === 0 && args.arrayInPath(fullPath).pop() === true) {
                const d = bigIntToUint8Array(1n)
                if (op == Op.CALLDATALOAD) {
                  vm.stack.peek().data = d
                } else {
                  const mem_off = uint8ArrayToBigInt(r1.data)
                  vm.memory.get(mem_off).data = d
                }
              }

              const new_label = new Arg({ offset: new_off, path: fullPath, add_val: 0, and_mask: null })
              if (op == Op.CALLDATALOAD) {
                vm.stack.peek().label = new_label
              } else {
                const mem_off = uint8ArrayToBigInt(r1.data)
                vm.memory.get(mem_off).label = new_label
                args.setTname(path, offset, 'bytes', 10)
              }
            }
          } else {
            const off = uint8ArrayToBigInt(r0.data)
            if (off >= 4n && off < 131072n - 1024n) {
              // -1024: cut 'trustedForwarder'
              args.getOrCreate([Number(off) - 4])

              const new_label = new Arg({ offset: Number(off) - 4, path: [], add_val: 0, and_mask: null })
              if (op == Op.CALLDATALOAD) {
                vm.stack.peek().label = new_label
              } else {
                const mem_off = uint8ArrayToBigInt(r1.data)
                vm.memory.get(mem_off).label = new_label
              }
            }
          }
        }
        break

      case Op.ADD:
        {
          const [l0, l1] = [r0.label, r1.label]
          if (l0 instanceof Arg && l1 instanceof Arg) {
            args.markNotBool(l0.path, l0.offset)
            args.markNotBool(l1.path, l1.offset)

            vm.stack.peek().label =
              l0.path.length > l1.path.length
                ? new Arg({
                    offset: l0.offset,
                    path: l0.path,
                    add_val: l0.add_val + l1.add_val,
                    and_mask: l0.and_mask,
                  })
                : new Arg({
                    offset: l1.offset,
                    path: l1.path,
                    add_val: l0.add_val + l1.add_val,
                    and_mask: l1.and_mask,
                  })
          } else if (l0 instanceof Arg || l1 instanceof Arg) {
            const [r, otd] = l0 instanceof Arg ? [r0, r1.data] : [r1, r0.data]
            const rl = r.label

            args.markNotBool(rl.path, rl.offset)

            const ot_val = uint8ArrayToBigInt(otd)

            const E256M1 = (1n << 256n) - 1n

            if (
              rl.offset == 0 &&
              rl.add_val == 0 &&
              rl.path.length != 0 &&
              uint8ArrayToBigInt(r.data) === 0n &&
              ot_val == E256M1
            ) {
              vm.stack.peek().data = bigIntToUint8Array(0n)
            }
            const add = (ot_val + BigInt(rl.add_val)) & E256M1
            if (add < 1n << 32n) {
              vm.stack.peek().label = new Arg({ offset: rl.offset, path: rl.path, add_val: Number(add), and_mask: rl.and_mask })
            }
          }
        }
        break

      case Op.MUL:
      case Op.SHL:
        {
          const [l0, l1] = [r0.label, r1.label]

          if ((op === Op.MUL && (l0 instanceof Arg || l1 instanceof Arg)) || (op === Op.SHL && l1 instanceof Arg)) {
            const [rl, ot] = l1 instanceof Arg ? [l1, r0] : [l0, r1]

            args.markNotBool(rl.path, rl.offset)
            if (ot.label instanceof Arg) {
              args.markNotBool(ot.label.path, ot.label.offset)
            }
            if (rl.offset === 0 && rl.add_val === 0) {
              if (rl.path.length != 0) {
                let mult = uint8ArrayToBigInt(ot.data)
                if (op === Op.SHL) {
                  mult = 1n << mult
                }
                if (mult === 1n) {
                  args.setTname(rl.path, null, 'bytes', 10)
                } else if (mult == 2n) {
                  args.setTname(rl.path, null, 'string', 20)
                } else if (mult % 32n === 0n && 32n <= mult && mult <= 3200n) {
                  args.setInfo(rl.path, new InfoValArray(Number(mult / 32n)))

                  const shouldUpdate = (v) => v.offset == 0 && v.path == rl.path && v.add_val == 0

                  vm.stack.data.forEach((el) => {
                    if (el.label instanceof Arg && shouldUpdate(el.label)) {
                      el.data = bigIntToUint8Array(1n)
                    }
                  })

                  vm.memory.data.forEach((el) => {
                    if (el.label instanceof Arg && shouldUpdate(el.label[1])) {
                      el.data = bigIntToUint8Array(1n)
                    }
                  })

                  vm.stack.peek().data = ot.data // ==bigIntToUint8Array(mult)
                }
              }
            }
          }
        }
        break

      case Op.GT:
      case Op.LT:
        {
          const [rl, ot] = op === Op.LT ? [r1.label, r0] : [r0.label, r1]
          if (ot.label instanceof Arg) {
            args.markNotBool(ot.label.path, ot.label.offset)
          }
          if (rl instanceof Arg) {
            args.markNotBool(rl.path, rl.offset)
            if (rl.offset === 0 && rl.add_val === 0 && rl.and_mask === null) {
              // 0 < arr.len || arr.len > 0
              const v = uint8ArrayToBigInt(ot.data)
              if (v === 0n || v === 31n) {
                vm.stack.peek().data = bigIntToUint8Array(1n)
              }
            }
          }
        }
        break

      case Op.AND:
        {
          const [l0, l1] = [r0.label, r1.label]
          if (l0 instanceof Arg || l1 instanceof Arg) {
            const [rl, otd] = l0 instanceof Arg ? [l0, r1.data] : [l1, r0.data]

            const { path, offset, add_val } = rl
            args.markNotBool(path, offset)

            const mask = uint8ArrayToBigInt(otd)
            let t = andMaskToType(mask)
            if (t !== null) {
              args.setTname(path, offset, t, 5)
              vm.stack.peek().label = new Arg({ offset, path, add_val, and_mask: mask })
            }
          }
        }
        break

      case Op.EQ:
        {
          const [l0, l1] = [r0.label, r1.label]
          if (l0 instanceof Arg && l1 instanceof Arg) {
            if (l0.offset === l1.offset && l0.path === l1.path && l0.add_val === l1.add_val) {
              let mask = null
              if (l0.and_mask === null && l1.and_mask != null) {
                mask = l1.and_mask
              } else if (l0.and_mask != null && l1.and_mask === null) {
                mask = l0.and_mask
              }
              if (mask !== null) {
                let t = andMaskToType(mask)
                if (t !== null) {
                  args.setTname(l0.path, l0.offset, t, 20)
                }
              }
            }
          }
        }
        break

      case Op.ISZERO:
        {
          const rl = r0.label
          if (rl instanceof Arg) {
            vm.stack.peek().label = new IsZeroResult(rl)
          } else if (rl instanceof IsZeroResult) {
            // Detect check for 0 in DIV, it's not bool in that case: ISZERO, ISZERO, PUSH off, JUMPI, JUMPDEST, DIV
            let is_bool = true
            const op = vm.code[vm.pc]
            if (op >= Op.PUSH1 && op <= Op.PUSH4) {
              const n = op - Op.PUSH0
              if (vm.code[vm.pc + n + 1] === Op.JUMPI) {
                const jumpdest = vm.code.subarray(vm.pc + 1, vm.pc + 1 + n).reduce((acc, b) => acc * 256 + b, 0)
                if (jumpdest + 1 < vm.code.length && vm.code[jumpdest] === Op.JUMPDEST && vm.code[jumpdest + 1] === Op.DIV) {
                  is_bool = false
                }
              }
            }
            if (is_bool) {
              args.setTname(rl.path, rl.offset, 'bool', 5)
            }
          }
        }
        break

      case Op.SIGNEXTEND:
        {
          const rl = r1.label
          if (rl instanceof Arg && r0 < 32n) {
            args.setTname(rl.path, rl.offset, `int${(Number(r0) + 1) * 8}`, 20)
          }
        }
        break

      case Op.BYTE:
        {
          const rl = r1.label
          if (rl instanceof Arg) {
            args.setTname(rl.path, rl.offset, 'bytes32', 4)
          }
        }
        break
    }
  }

  return args.joinToString()
}

File: evmole/arguments.py
from dataclasses import dataclass, field

from .evm.element import Element
from .evm.opcodes import Op
from .evm.stack import StackIndexError
from .evm.vm import UnsupportedOpError, Vm
from .utils import to_bytes


@dataclass(frozen=True)
class Arg:
    offset: int
    path: tuple[int, ...]
    add_val: int
    and_mask: int | None


@dataclass(frozen=True)
class IsZeroResult:
    offset: int
    path: tuple[int, ...]
    add_val: int
    and_mask: int | None


@dataclass(frozen=True)
class InfoValDynamic:
    n_elements: int


@dataclass(frozen=True)
class InfoValArray:
    n_elements: int


@dataclass
class Info:
    tinfo: InfoValDynamic | InfoValArray | None = None
    tname: tuple[str, int] | None = None
    children: dict[int, 'Info'] = field(default_factory=dict)

    def to_str(self, is_root: bool = False) -> str:
        if self.tname is not None:
            name = self.tname[0]
            if name == 'bytes':
                if (
                    self.tinfo is None
                    or (isinstance(self.tinfo, InfoValArray) and self.tinfo.n_elements == 0)
                    or (isinstance(self.tinfo, InfoValDynamic) and self.tinfo.n_elements == 1)
                ):
                    return name
            elif len(self.children) == 0:
                if self.tinfo is None or isinstance(self.tinfo, InfoValDynamic):
                    return name

        start_key = 32 if isinstance(self.tinfo, InfoValArray) else 0
        end_key = max(self.children.keys()) if self.children else 0
        if isinstance(self.tinfo, (InfoValArray, InfoValDynamic)):
            end_key = max(end_key, self.tinfo.n_elements * 32)

        q = []
        for k in range(start_key, end_key + 1, 32):
            q.append(self.children[k].to_str(False) if k in self.children else 'uint256')

        c = f'({",".join(q)})' if len(q) > 1 and not is_root else ','.join(q)

        if isinstance(self.tinfo, InfoValArray):
            return f'{c}[]'

        if isinstance(self.tinfo, InfoValDynamic):
            if end_key == 0 and not self.children:
                return 'bytes'
            if end_key == 32:
                if not self.children:
                    return 'uint256[]'
                if len(self.children) == 1 and next(iter(self.children.values())).tinfo is None:
                    return f'{q[1]}[]'

        return c


class ArgsResult:
    data: Info
    not_bool: set[tuple[int, ...]]

    def __init__(self):
        self.data = Info()
        self.not_bool = set()

    def get_or_create(self, path: tuple[int, ...]):
        node = self.data
        for key in path:
            if key not in node.children:
                node.children[key] = Info()
            node = node.children[key]
        return node

    def get(self, path: tuple[int, ...]):
        node = self.data
        for key in path:
            if key not in node.children:
                return None
            node = node.children[key]
        return node

    def mark_not_bool(self, path: tuple[int, ...], offset: int):
        full_path = (*path, offset)
        el = self.get(full_path)
        if el and el.tname and el.tname[0] == 'bool':
            el.tname = None
        self.not_bool.add(full_path)

    def set_tname(self, path: tuple[int, ...], offset: int | None, tname: str, confidence: int):
        full_path = (*path, offset) if offset is not None else path
        if tname == 'bool' and full_path in self.not_bool:
            return
        el = self.get_or_create(full_path)
        if el.tname is not None and confidence <= el.tname[1]:
            return
        el.tname = (tname, confidence)

    def array_in_path(self, path: tuple[int, ...]):
        ret = []
        el: Info | None = self.data
        for p in path:
            if el is not None:
                el = el.children.get(p)
                ret.append(el and isinstance(el.tinfo, InfoValArray))
            else:
                ret.append(False)
        return ret

    def set_info(self, path: tuple[int, ...], tinfo: InfoValDynamic | InfoValArray):
        if len(path) == 0:  # root
            return
        el = self.get_or_create(path)
        if isinstance(tinfo, InfoValDynamic):
            if isinstance(el.tinfo, InfoValDynamic) and el.tinfo.n_elements > tinfo.n_elements:
                return
            if isinstance(el.tinfo, InfoValArray):
                return
        if isinstance(el.tinfo, InfoValArray) and isinstance(tinfo, InfoValArray):
            if tinfo.n_elements < el.tinfo.n_elements:
                return
        el.tinfo = tinfo

    def join_to_string(self):
        return '' if not self.data.children else self.data.to_str(True)


def and_mask_to_type(mask: int) -> str | None:
    if mask == 0:
        return None
    if (mask & (mask + 1)) == 0:
        # 0x0000ffff
        bl = mask.bit_length()
        if bl % 8 == 0:
            return 'address' if bl == 160 else f'uint{bl}'
    else:
        # 0xffff0000
        m = int.from_bytes(mask.to_bytes(32, 'big'), 'little')
        if (m & (m + 1)) == 0:
            bl = m.bit_length()
            if bl % 8 == 0:
                return f'bytes{bl // 8}'
    return None


def function_arguments(code: bytes | str, selector: bytes | str, gas_limit: int = int(5e4)) -> str:
    bytes_selector = to_bytes(selector)
    vm = Vm(code=to_bytes(code), calldata=Element(data=bytes_selector, label='calldata'))
    gas_used = 0
    inside_function = False
    args = ArgsResult()
    while not vm.stopped:
        try:
            if inside_function:
                # print(vm, '\n', sep='')
                # print(args.data)
                # print(args.join_to_string())
                pass
            ret = vm.step()
            gas_used += ret[1]
            if gas_used > gas_limit:
                # raise Exception(f'gas overflow: {gas_used} > {gas_limit}')
                break
        except (StackIndexError, UnsupportedOpError) as ex:
            _ = ex
            # print(ex)
            break

        if inside_function is False:
            if ret[0] in {Op.EQ, Op.XOR, Op.SUB}:
                p = int.from_bytes(vm.stack.peek().data, 'big')
                if p == (1 if ret[0] == Op.EQ else 0):
                    inside_function = ret[2].data.endswith(bytes_selector)
            continue

        match ret:
            case (Op.CALLDATASIZE, _):
                vm.stack.pop()
                vm.stack.push_uint(131072)

            case ((Op.CALLDATALOAD | Op.CALLDATACOPY) as op, _, Element(Arg(offset, path, add_val, _)), *sa):
                if add_val >= 4 and (add_val - 4) % 32 == 0:
                    full_path = (*path, offset)
                    po = 0
                    if add_val != 4:
                        po += sum(32 if is_arr else 0 for is_arr in args.array_in_path(path))
                        if po > (add_val - 4):
                            po = 0
                    new_off = add_val - 4 - po
                    args.set_info(full_path, InfoValDynamic(new_off // 32))

                    mem_offset = int.from_bytes(sa[0].data, 'big') if op == Op.CALLDATACOPY else 0

                    if new_off == 0 and args.array_in_path(full_path)[-1]:
                        if op == Op.CALLDATALOAD:
                            vm.stack.peek().data = (1).to_bytes(32, 'big')
                        else:
                            m = vm.memory.get(mem_offset)
                            if m is not None:
                                m.data = (1).to_bytes(32, 'big')

                    new_label = Arg(new_off, full_path, 0, None)
                    if op == Op.CALLDATALOAD:
                        vm.stack.peek().label = new_label
                    else:
                        args.set_tname(path, offset, 'bytes', 10)
                        m = vm.memory.get(mem_offset)
                        if m is not None:
                            m.label = new_label

            case ((Op.CALLDATALOAD | Op.CALLDATACOPY) as op, _, Element() as el, *sa):
                off = int.from_bytes(el.data, 'big')
                if 4 <= off < 131072 - 1024:  # -1024: cut 'trustedForwarder'
                    args.get_or_create((off - 4,))
                    new_label = Arg(off - 4, (), 0, None)
                    if op == Op.CALLDATALOAD:
                        vm.stack.peek().label = new_label
                    else:
                        mem_offset = int.from_bytes(sa[0].data, 'big')
                        m = vm.memory.get(mem_offset)
                        if m is not None:
                            m.label = new_label

            case (
                Op.ADD,
                _,
                Element(Arg(f_offset, f_path, f_add_val, f_and_mask)),
                Element(Arg(s_offset, s_path, s_add_val, s_and_mask)),
            ):
                args.mark_not_bool(f_path, f_offset)
                args.mark_not_bool(s_path, s_offset)
                p = vm.stack.peek()
                if len(f_path) > len(s_path):
                    p.label = Arg(f_offset, f_path, f_add_val + s_add_val, f_and_mask)
                else:
                    p.label = Arg(s_offset, s_path, f_add_val + s_add_val, s_and_mask)

            case (
                (Op.ADD, _, Element(Arg(offset, path, add_val, and_mask)) as el, Element() as ot)
                | (Op.ADD, _, Element() as ot, Element(Arg(offset, path, add_val, and_mask)) as el)
            ):
                args.mark_not_bool(path, offset)

                ot_val = int.from_bytes(ot.data, 'big')

                E256M1 = (1 << 256) - 1
                if (
                    offset == 0
                    and add_val == 0
                    and len(path) != 0
                    and int.from_bytes(el.data, 'big') == 0
                    and int.from_bytes(ot.data, 'big') == E256M1
                ):
                    vm.stack.peek().data = (0).to_bytes(32, 'big')

                add = (ot_val + add_val) & E256M1
                if add < (1 << 32):
                    vm.stack.peek().label = Arg(offset, path, add, and_mask)

            case (
                (Op.MUL as op, _, Element(Arg(0, path, 0, _)) as el, Element() as ot)
                | (Op.MUL as op, _, Element() as ot, Element(Arg(0, path, 0, _)) as el)
                | (Op.SHL as op, _, Element() as ot, Element(Arg(0, path, 0, _)) as el)
            ):
                args.mark_not_bool(path, 0)
                if isinstance(ot.label, Arg):
                    args.mark_not_bool(ot.label.path, ot.label.offset)
                if len(path) != 0:
                    mult = int.from_bytes(ot.data, 'big')
                    if op == Op.SHL:
                        mult = 1 << mult

                    if mult == 1:
                        args.set_tname(path, None, 'bytes', 10)
                    elif mult == 2:
                        args.set_tname(path, None, 'string', 20)
                    elif mult % 32 == 0 and 32 <= mult <= 3200:
                        args.set_info(path, InfoValArray(mult // 32))

                        for el in vm.stack.data:
                            if isinstance(el.label, Arg):
                                lab = el.label
                                if lab.offset == 0 and lab.path == path and lab.add_val == 0:
                                    el.data = (1).to_bytes(32, 'big')

                        for el in vm.memory.data:
                            if isinstance(el[1].label, Arg):
                                lab = el[1].label
                                if lab.offset == 0 and lab.path == path and lab.add_val == 0:
                                    el[1].data = (1).to_bytes(32, 'big')

                        vm.stack.peek().data = ot.data  # = mult

            case (
                (Op.GT, _, Element(Arg(0, path, 0, None)), Element() as ot)
                | (Op.LT, _, Element() as ot, Element(Arg(0, path, 0, None)))
            ):
                args.mark_not_bool(path, 0)
                v = int.from_bytes(ot.data, 'big')
                if v == 0 or v == 31:
                    vm.stack.peek().data = (1).to_bytes(32, 'big')

            case (
                (Op.LT | Op.GT | Op.MUL, _, Element(Arg(offset, path, _, _)), _)
                | (Op.LT | Op.GT | Op.MUL, _, _, Element(Arg(offset, path, _, _)))
            ):
                args.mark_not_bool(path, offset)

            case (
                (Op.AND, _, Element(Arg(offset, path, add_val, None)), Element() as ot)
                | (Op.AND, _, Element() as ot, Element(Arg(offset, path, add_val, None)))
            ):
                args.mark_not_bool(path, offset)
                mask = int.from_bytes(ot.data, 'big')
                t = and_mask_to_type(mask)
                if t is not None:
                    args.set_tname(path, offset, t, 5)
                    vm.stack.peek().label = Arg(offset, path, add_val, mask)

            case (
                (Op.EQ, _, Element(Arg(offset, path, add_val, None)), Element(Arg(s_offset, s_path, s_add_val, mask)))
                | (Op.EQ, _, Element(Arg(s_offset, s_path, s_add_val, mask)), Element(Arg(offset, path, add_val, None)))
            ):
                if s_offset == offset and s_path == path and s_add_val == add_val and mask is not None:
                    t = and_mask_to_type(mask)
                    if t is not None:
                        args.set_tname(path, offset, t, 20)

            case (Op.ISZERO, _, Element(Arg(offset, path, add_val, and_mask))):
                vm.stack.peek().label = IsZeroResult(offset, path, add_val, and_mask)

            case (Op.ISZERO, _, Element(IsZeroResult(offset, path, _, _))):
                # Detect check for 0 in DIV, it's not bool in that case: ISZERO, ISZERO, PUSH off, JUMPI, JUMPDEST, DIV
                is_bool = True
                op = vm.code[vm.pc]
                if op >= Op.PUSH1 and op <= Op.PUSH4:
                    n = op - Op.PUSH0
                    if vm.code[vm.pc + n + 1] == Op.JUMPI:
                        jumpdest = int.from_bytes(vm.code[(vm.pc + 1) : (vm.pc + 1 + n)], signed=False)
                        if jumpdest + 1 < len(vm.code) and vm.code[jumpdest] == Op.JUMPDEST and vm.code[jumpdest + 1] == Op.DIV:
                            is_bool = False
                if is_bool:
                    args.set_tname(path, offset, 'bool', 5)

            case (Op.SIGNEXTEND, _, s0, Element(Arg(offset, path, _, _))):
                if s0 < 32:
                    args.set_tname(path, offset, f'int{(s0+1)*8}', 20)

            case (Op.BYTE, _, _, Element(Arg(offset, path, add_val, and_mask))):
                args.set_tname(path, offset, 'bytes32', 4)

    return args.join_to_string()
